### ISSUE TITLE
Automatic Binning estimation for ECE and Brier Score Metric

### ISSUE THREAD

**Issue by lorenzofamiglini** [NannyML/nannyml#92]
It would be useful to exploit methodologies that allow the user to avoid the choice of the number of bins for estimating the Expected Calibration Error. From the literature, static binning without observing the density of the confidence scores presents several bottlenecks. Also, in this case, we need to take into account the trade-off between variance and bias. It was observed that a fair number of bins to be selected is in the range of [10,20]. Nevertheless, manual binning could be misleading. 

My solution is the following: 

We can exploit different approaches already implemented in the Numpy framework: Freedman Diaconis Estimator, where the span (width) of each bin is proportional to the interquartile range of the confidence score distribution and inversely proportional to the cube root of cs size. 
Depending on the dataset size  estimate the number of bins. You can find everything of this in the following link: 

https://numpy.org/doc/stable/reference/generated/numpy.histogram_bin_edges.html

**If we want to avoid binning issues: Brier Score** 

Brier Score is a strictly proper score function that can measure the calibration without discretising the confidence scores space into bins. It can be used either for binary or multiclass classification: the formulation for multiclass, even if it does not maintain the same properties of binary form, it is still a robust metric. Please, pay attention that if the multiclass version is not divided by 2 x N, the boundaries will be 0 and 2. 

Hope this helps! 

Cheers

**Comment by lorenzofamiglini** [comment]
An example of how we can compute ECE (for binary settings, but it can be easly extended to the multiclass ones) with automatic binning is the following: 

```

import numpy as np 

def binning_schema(prob, method = 'auto', ndim = 1):
    """
    ndim: which column should be selected for the proba vector: 
          0 prob for class 0, 1 prob for class 1 proba
    """
    prob = prob[:,ndim]
    
    b = np.histogram_bin_edges(prob, bins=method, weights=None)
    
    binids = np.digitize(prob, b) - 1
    relative_freq_bin = np.unique(binids, return_counts=True)[1] / len(prob)
    return {'bins': b,
            'binids': binids,
            'binfr': relative_freq_bin}

def compute_eces(y,pred, prob, binids, bins, ndim = 1):
    
    prob = prob[:,ndim]
    bin_total = np.bincount(binids, minlength=len(bins))
    nonzero = bin_total != 0
    
    bin_true = np.bincount(binids, weights=y, minlength=len(bins))
    prob_true = bin_true[nonzero] / bin_total[nonzero]
    confscore_bins = np.bincount(binids, weights=prob)[nonzero]/bin_total[nonzero]    
    loce_w = bin_total[nonzero]/len(y)    
    lece = np.abs(prob_true - confscore_bins)
    ece = np.sum(loce_w*lece)

    return ece, lece

```

**Comment by lorenzofamiglini** [comment]
While for Brier Score in general settings, both binary and multiclass (this implementation is in PyTorch, but it can be replaced with Numpy): 

```

def brierscore(input, target): 
    target = target.view(-1,1)
    y_one_hot = torch.nn.functional.one_hot(target,num_classes=input.shape[1])
    
    pt = F.softmax(input, dim = 1)
    squared_diff = torch.sum((y_one_hot.squeeze() - pt) ** 2, axis = 1)
    sum_squared_diff = torch.sum(squared_diff)
    loss = sum_squared_diff / (float(2*pt.shape[0])) #*2 for boundaries between [0,1]

    return loss
```

**Comment by nikml** [comment]
An update regarding your suggestions @lorenzofamiglini 

I can see now using [np.histogram_bin_edges](https://numpy.org/doc/stable/reference/generated/numpy.histogram_bin_edges.html) will be an improvement over [`_get_bin_index_edges`](https://github.com/NannyML/nannyml/blob/72b1aad79c08ad5942280b35f7ba6398587e566d/nannyml/calibration.py#L152-L187) and will bring it up to the team for consideration and prioritization.

Regarding the Brier Score, I like it and think it's also better than binning. However I 'd like to think a little bit more about it with regards to how it interacts with calicration, especially multiclass. We may need to test the effects on some datasets to see if the results differ, and if so how.

We 'll be posting here once I have more to share.

**Comment by lorenzofamiglini** [comment]
Happy to hear that! 

Yes, my suggest is to evaluate calibration by exploiting also statistical tests such as https://en.m.wikipedia.org/wiki/Hosmer%E2%80%93Lemeshow_test. Brier Score is not sensitive to the binning scheme (indeed, in this case every confidence score is a single bin), and it has several strong properties related to proper scoring rules. 

A good assessment for ECE (please, remember to use the fraction of positive and not the accuracy is not a good generalisation) could be to evaluate the standard deviation of ECE by changing either the binning method or the number of bins.

**Comment by nnansters** [comment]
Update: we've received a great PR (#172) by @Jebq that incorporates the `numpy.historgram_bin_edges` functionality.
